src/Trains.NET.Rendering/Game/Game.cs

```diff
@@ -148,30 +143,23 @@ public void DrawFrame()
 
             IPixelMapper? pixelMapper = _pixelMapper.Snapshot();
 
-            using IImageCanvas? imageCanvas = _imageFactory.CreateImageCanvas(_width, _height);
+            using IImageCanvas imageCanvas = _imageFactory.CreateImageCanvas(_width, _height);
 
             ICanvas? canvas = imageCanvas.Canvas;
             RenderFrame(canvas, pixelMapper);
 
-            IImage? oldBuffer = _backBuffer;
-            lock (_bufferLock)
-            {
-                _backBuffer = imageCanvas.Render();
-            }
-            oldBuffer?.Dispose();
+            _imageCache.Set(this, imageCanvas.Render());
 
-            if (_needsBufferReset)
+            foreach (IScreen? screen in _screens)
             {
-                _gameBufferReset.Start();
-
-                foreach (IImage image in _imageBuffer.Values)
+                if (_imageCache.IsDirty(screen))
                 {
-                    image.Dispose();
+                    _screenDrawTimes[screen].Start();
+                    using IImageCanvas? screnCanvas = _imageFactory.CreateImageCanvas(_screenWidth, _screenHeight);
+                    screen.Render(screnCanvas.Canvas, _screenWidth, _screenHeight);
+                    _screenDrawTimes[screen].Stop();
+                    _imageCache.Set(screen, screnCanvas.Render());
```

Comment:
You have a race condition here, this Set call can dispose the Image after Get is called, but before it is drawn.

Reply:
In reality yes, in practice this already existed with the back buffer and didn't cause any problems. Also since what happens with the return of Get is external to the image cache there isn't a great way to deal with it, except perhaps keeping more images in the cache, and disposing after the 2nd set. 

Nothing should be doing anything with the result of Get except immediately rendering it and throwing it away anyway.

Reply:
Note to future self: Scott reported getting AVs so have a better think about this!!